### ISSUE TITLE
Generic Grain Methods

### ISSUE THREAD

**Issue by Joshua-Ferguson** [dotnet/orleans#1352]
I created an Orleans grain with a generic method call and I noticed that codegen failed. I'm not sure if it should fail on generics (intentional for performance reasons, or user error?). I created a [PR](https://github.com/Joshua-Ferguson/orleans/pull/1) on a fork to show the changes I made to get generics working. But I'm not sure the changes are made are the best way to do it (really just wanted to be able to test that it could be done). I got the type info < T > passed where I could and used < dynamic > everywhere else.

```
public interface ISecretNinjaTestGrain : IGrainWithIntegerKey
{
    Task<int> FakeGeneric<T>();
    Task<int> GenericArg<T>(T data);
    Task<T> SimpleGeneric<T>();
    Task<int> SimpleGenericInteger<T>();
    Task<int> SimpleInteger();
    Task<T> VeryGeneric<T>(T data);
    Task<string> MultiGeneric<T, T2, T3>(T t, T2 t2, T3 t3);
    Task<T2> BreakDance<T, T2, T3>(T t, T2 t2, T3 t3);
    Task<T> LessSimpleGeneric<T>(T one, T two);
}

public class SecretNinjaTestGrain : Grain, ISecretNinjaTestGrain
{
    public Task<int> SimpleInteger() => Task.FromResult(3);
    public Task<int> SimpleGenericInteger<T>() => Task.FromResult(3);
    public Task<T> SimpleGeneric<T>() => Task.FromResult(default(T));
    public Task<T> LessSimpleGeneric<T>(T one, T two) => Task.FromResult(default(T));
    public Task<int> FakeGeneric<T>() => Task.FromResult(3);
    public Task<int> GenericArg<TDifferent>(TDifferent data) => Task.FromResult(3);
    public Task<T> VeryGeneric<T>(T data) => Task.FromResult(data);
    public Task<string> MultiGeneric<T, T2, T3>(T t, T2 t2, T3 t3) => Task.FromResult($"{t}{t2}{t3}");
    public Task<T2> BreakDance<T, T2, T3>(T t, T2 t2, T3 t3) => Task.FromResult(default(T2));
}
```

I was able to make calls to all these methods.

**Comment by Joshua-Ferguson** [comment]
I was a little premature. The test I wrote only looked at the Task object for LessSimpleGeneric and BreakDance and not the return value, so they throw a null object exception if you try to get the Result. But MultiGeneric works just fine.

**Comment by ReubenBond** [comment]
Thanks, @Joshua-Ferguson! Could you include your tests in the PR? I'd like to check those out, too. I think this is a viable approach to supporting generic methods; we can work through the edge cases.

**Comment by sergeybykov** [comment]
I'm pretty sure we do not support generic methods on non-generic grain interfaces/classes. 

Today, generic arguments are taken at activation time, and all supporting objects, such as invoker, are instantiated and cached at that point with those type arguments. A method with a type argument independent from the grain class type argument (of if the class is not generic) requires instantiation of per-method invokers for each argument type, potentially different for every call. It's a feasible thing to implement, just not there yet.

A potential workaround here is to declare `ISecretNinjaTestGrain` as `ISecretNinjaTestGrain<T, T2, T3>`. The price for that will be inability to invoke generic methods with type arguments different from those passes to `GrainFactory.GetGrain<ISecretNinjaTestGrain<U1, U2, U3>>`

**Comment by ReubenBond** [comment]
@sergeybykov @Joshua-Ferguson's approach is to use `dynamic` parameters for generic methods. Do you think this will cause a problem?

**Comment by sergeybykov** [comment]
Oh, sorry, I missed the PR part and replied about the current behavior. Will try to comprehend the PR.

**Comment by Joshua-Ferguson** [comment]
@ReubenBond I've been working on this as I get time. What I found out, was that dynamic only works for value types. Once you go to classes and structs it fails.

Its kind of like this:
```
method<int>(3) --> GenericTypeArguments --> System.Int; 
method<dynamic>(3) --> GenericTypeArguments --> System.Obj;
```

Once the Generic Type Arguments become objects they are lost (I feel like they shouldn't be, but If they aren't, I'm not sure what the correct approach is to retrieve them -- once we have them its easy to reconstruct the object in the deserializer). This isn't a problem for value types, but it is a problem for classes and structs.

okay: `Task<string> Foo<T, T2, T3>(T t, T2 t2, T3 t3) => Task.FromResult($"{t}{t2}{t3}");`
okay: `Task<T> Foo<T>(T data) => Task.FromResult(data);`
bad: 
```
Task<KeyValuePair<TKey, TValue>> FooKeyValuePair<TKey, TValue>(TKey key, TValue value)
{
    return Task.FromResult(new KeyValuePair<TKey, TValue>(key, value));
}
```

I'm thinking the best solution to reconstruct type parameters, would be to pass the generic type information along with the object[] arguments.

```
public global::System.Threading.Tasks.Task<T> @Generic<T>(T @one, T @two)
{
    return base.@InvokeMethodAsync<T>(1639928696, new global::System.Object[]{@one, @two});
}
```
Would become 
```
public global::System.Threading.Tasks.Task<T> @Generic<T>(T @one, T @two)
{
    return base.@InvokeMethodAsync<T>(1639928696, new global::System.Object[]{@one, @two}, new global::System.Type[]{typeof(T)});
}
```

What do you think?

**Comment by Joshua-Ferguson** [comment]
I have updated the PR. Dynamic only works for parameters that are made of non constructed generic types. Constructed generic types are now supported.

Now you can do things like:
```
var dict = new Dictionary<int, string>();
var dict2 = await ninjaGrain.DropFirstEntry(dict);
var dict3 = await ninjaGrain.DropEntry(dict2, 1);
```
@ReubenBond Test Project is in the PR comments. Apologies for the mess.

**Comment by creyke** [comment]
+1 for this functionality. Hugely useful.

**Comment by Joshua-Ferguson** [comment]
I've updated this functionality for the patch, you can find it [here](https://github.com/Joshua-Ferguson/orleans/commit/4b2197cb692d478c1c2dfa3264110280c7f0df46).

I dropped the 'dynamic' path and now it uses just the GenericInvoker class which simplifies the changes I had to make to the OrleansCodeGenerator by a lot. I've added unit tests. I fixed a lot of the Roslyn code gen things that worked, but maybe won't quite as right (now looks more like @ReubenBond style code).

I swapped out Type for TypeInfo nearly everywhere (I believe that's the right move for dot net core compatibility?).

There is still one use of dynamic as the return type of the GenericInvoker.Invoke method. I've experimented with other options, but I haven't found anything that works yet. Which means the Microsoft.CSharp dll needs to be referenced in the test project.

Should I attempt to submit this as a PR? (git skills improving, still got miles to go, plus with the amount of changes this code implements, I still fear I am missing things)

**Comment by ReubenBond** [comment]
Thanks, @Joshua-Ferguson! I've added my comments to https://github.com/Joshua-Ferguson/orleans/commit/4b2197cb692d478c1c2dfa3264110280c7f0df46

**Comment by Joshua-Ferguson** [comment]
Thanks, @ReubenBond ! I have updated the code based on your comments. 
https://github.com/Joshua-Ferguson/orleans/commit/cf5b68d050222170398b42b4b4b0fffe53f517ca and https://github.com/Joshua-Ferguson/orleans/commit/d2248f4df76d40b6764ab4454e19f08ceedd9ff7 (very small).

**Comment by ReubenBond** [comment]
@Joshua-Ferguson I submitted a pull request against your fork to update it to current head so that we can work on this. The PR is a bit dated not, but if you could rebase your changes on current master, that would be helpful. Let me know if you run into issues

**Comment by sergeybykov** [comment]
Where are we currently at with this?

**Comment by sergeybykov** [comment]
Resolved via #2670.
